server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java

```diff
@@ -0,0 +1,257 @@
+     * The time at which the cluster state update began that found a shutdown signal for this node. Negative value means unset (node is not
+     * shutting down).
+     */
+    private volatile long shutdownStartMillis = -1;
+
+    /**
+     * The time at which the cluster state finished setting shard snapshot states to PAUSING, which the shard snapshot operations will
+     * discover asynchronously. Negative value means unset (node is not shutting down)
+     */
+    private volatile long shutdownFinishedSignallingPausingMillis = -1;
+
+    /**
+     * Tracks the number of shard snapshots that have started on the data node but not yet finished.
+     */
+    private final AtomicLong numberOfShardSnapshotsInProgressOnDataNode = new AtomicLong();
+
+    /**
+     * The logic to track shard snapshot status update requests to master can result in duplicate requests (see
+     * {@link ResultDeduplicator}), as well as resending requests if the elected master changes.
+     * Tracking specific requests uniquely by snapshot ID + shard ID de-duplicates requests for tracking.
+     */
+    private final Set<String> shardSnapshotRequests = ConcurrentCollections.newConcurrentSet();
+
+    /**
+     * Track how the shard snapshots reach completion during shutdown: did they fail, succeed or pause?
+     */
+    private final AtomicLong doneCount = new AtomicLong();
+    private final AtomicLong failureCount = new AtomicLong();
+    private final AtomicLong abortedCount = new AtomicLong();
+    private final AtomicLong pausedCount = new AtomicLong();
+
+    public SnapshotShutdownProgressTracker(ClusterService clusterService, Settings settings, ThreadPool threadPool) {
+        this.clusterService = clusterService;
+        this.progressLoggerInterval = SNAPSHOT_PROGRESS_DURING_SHUTDOWN_INTERVAL_TIME_SETTING.get(settings);
+        this.threadPool = threadPool;
+    }
+
+    private void scheduleProgressLogger() {
+        scheduledProgressLoggerFuture = threadPool.scheduleWithFixedDelay(
+            progressLogger,
+            progressLoggerInterval,
+            threadPool.executor(ThreadPool.Names.GENERIC)
+        );
+        logger.trace(
+            Strings.format(
+                "Starting shutdown snapshot progress logging on node [%s], runs every [%s]",
+                clusterService.state().nodes().getLocalNodeId(),
+                progressLoggerInterval
+            )
+        );
+    }
+
+    private void cancelProgressLogger() {
+        assert scheduledProgressLoggerFuture != null : "Somehow shutdown mode was removed before it was added.";
+        scheduledProgressLoggerFuture.cancel();
+        logger.trace(
+            Strings.format("Cancelling shutdown snapshot progress logging on node [%s]", clusterService.state().nodes().getLocalNodeId())
+        );
+    }
+
+    /**
+     * Logs some statistics about shard snapshot progress.
+     */
+    private void logProgressReport() {
+        assert this.shutdownStartMillis != -1;
```

Comment:
I think this might not always be true, if we're cancelled on another thread just before executing this line.